catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java

```diff
@@ -1644,6 +1688,237 @@ public void setMetacardTypes(List<MetacardType> types) {
+        transactionRequest.setVerbose(true);
+
+        List<Metacard> metacards = createRequest.getMetacards();
+        List<String> metacardIds = metacards.stream()
+                .map(mcard -> mcard.getId())
+                .collect(Collectors.toList());
+        ArrayList<Metacard> createdMetacards = new ArrayList<>();
+        ArrayList<Filter> createdMetacardFilters = new ArrayList<>();
+        HashSet<ProcessingDetails> errors = new HashSet<>();
+
+        String insertTypeName = schemaTransformerManager.getTransformerIdForSchema(
+                cswSourceConfiguration.getOutputSchema());
+
+        if (insertTypeName == null) {
+            insertTypeName = CswConstants.CSW_RECORD;
+        }
+
+        transactionRequest.getInsertActions()
+                .add(new InsertAction(insertTypeName, null, metacards));
+        try {
+            TransactionResponseType response = csw.transaction(transactionRequest);
+            Set<String> processedIds = new HashSet<>();
+            //dive down into the response to get the created ID's. We need these so we can query
+            //the source again to get the created metacards and put them in the result
+            createdMetacardFilters.addAll(response.getInsertResult()
+                    .stream()
+                    .map(InsertResultType::getBriefRecord)
+                    .flatMap(Collection::stream)
+                    .map(BriefRecordType::getIdentifier)
+                    .flatMap(Collection::stream)
+                    .map(JAXBElement::getValue)
+                    .map(SimpleLiteral::getContent)
+                    .flatMap(Collection::stream)
+                    .map(id -> {
+                        processedIds.add(id);
+                        return filterBuilder.attribute(Metacard.ID)
+                                .is()
+                                .equalTo()
+                                .text(id);
+                    })
+                    .collect(Collectors.toList()));
+            metacardIds.removeAll(processedIds);
+
+            for (String id : metacardIds) {
+                errors.add(new ProcessingDetailsImpl(id, null, "Failed to create metacard"));
+            }
+
+        } catch (CswException e) {
+            throw new IngestException("Csw Transaction Failed : " + e.getMessage());
+        }
+
+        try {
+            createdMetacards.addAll(transactionQuery(createdMetacardFilters));
+        } catch (UnsupportedQueryException e) {
+            errors.add(new ProcessingDetailsImpl(this.getId(),
+                    e,
+                    "Failed to retrieve newly created metacards"));
+        }
+
+        return new CreateResponseImpl(createRequest, properties, createdMetacards, errors);
+    }
+
+    @Override
+    public UpdateResponse update(UpdateRequest updateRequest) throws IngestException {
+        Map<String, Serializable> properties = new HashMap<>();
+
+        Subject subject =
```

Comment:
Creating & initializing the TransactionRequest is duplicated in multiple places - move to a common method.